### DISCUSSION TITLE
Mouse tracking for dyadic experiment

### DISCUSSION THREAD

**Discussion by HuangHam** [empiricaly/empirica#530] [Q&A]
I am interested in writing up an experiment where there will only be 2 players. At one stage, I want each player to view the mouse of the other player in real-time. If this becomes intractable, it would also suffice to show which region the other player's mouse is currently in (out of 36 regions on the screen). What I am currently doing is in the callback.js file, I added in Empirica.onStageStart for (const player of players) {
  player.set('mouseCoord', [x, y])
  } and do partner.get('mouseCoord') in the stage file. However, this value is undefined when I test it on local host.

**Comment by npaton** [comment]
I am not sure about the mouseCoord being undefined here, setting a value in onStageStart and seeing on the client side is a pretty fundamental feature that should just works, but your exact situation might be different then what I am imagining. If you wouldn't mind, please create a ticket for that with all the code details of your situation, and feel free to share with me the code in DM on Slack if you want me to take a closer look.

As for the mouse tracking, the last release added what I'm calling `ephemeral` attribute updates (not persisted to the database). This would be ideal for mouse tracking, since persisting mouse mouvement could get hairy for the database (see bullet point #1 below). If you're ok not persisting the mouse movements, then I'd recommend using this feature. And maybe if you still want some persistence, you can use the regions idea for slightly less granular position info, which should be a lot less data.

Check out the [release notes](https://github.com/empiricaly/empirica/releases/tag/%40empirica%2Fcore%401.10.0) to learn how to use ephemeral attributes.

I'll post here my answer about mouse tracking in Slack for future reference, for anyone else wanting to do this:

> There are a few issues with mouse tracking:
> - It will produce a large amount of data. Depending on how many participants, how active they are, and how long the games are, it could be overwhelming for the system. (NOTE: this can be improved with the newly release ephemeral attributes, see above)
> - `empirica export` currently does not output all the versions of a key. For example, you record the mouse in player.set("mouse", mouseposition) , the export will only return the last version of the "mouse" value. But you can parse the tajriba.json file to get back all the previous versions if you need to, so there is a way.
> - Naive mouse tracking might be done relative to the page rendered, from the top left corner of the screen. If the UI is dynamic (e.g. centered), the positions will not correlate with UI elements. One solution here is to use a fixed layout (fixed width), that aligns to the top left of the screen. Another solution would be to have the UI be fixed within a "bounding box" and use the coordinates of that box.
